File: bestia-zoneserver/src/main/java/bestia/entity/component/interceptor/ActorUpdateComponentInterceptor.java
package bestia.entity.component.interceptor;

import bestia.entity.Entity;
import bestia.entity.EntityService;
import bestia.entity.component.Component;
import bestia.entity.component.ComponentActor;
import bestia.messages.ComponentChangedMessage;
import bestia.messages.ComponentMessageEnvelope;
import bestia.messages.MessageApi;

import java.util.Objects;

/**
 * This interceptor will check if the component was annotated to notify certain actors
 * if the component value has changed. The actor will then get notified.
 */
public class ActorUpdateComponentInterceptor extends BaseComponentInterceptor<Component> {

  private final MessageApi msgApi;

  public ActorUpdateComponentInterceptor(MessageApi msgApi) {
    super(Component.class);

    this.msgApi = Objects.requireNonNull(msgApi);
  }

  @Override
  protected void onDeleteAction(EntityService entityService, Entity entity, Component comp) {
    final ComponentActor syncActor = comp.getClass().getAnnotation(ComponentActor.class);
    if (dontUpdateActor(syncActor)) {
      return;
    }
    updateActorComponent(entity, comp);
  }

  @Override
  protected void onUpdateAction(EntityService entityService, Entity entity, Component comp) {
    final ComponentActor syncActor = comp.getClass().getAnnotation(ComponentActor.class);
    if (dontUpdateActor(syncActor)) {
      return;
    }
    updateActorComponent(entity, comp);
  }

  @Override
  protected void onCreateAction(EntityService entityService, Entity entity, Component comp) {
    final ComponentActor syncActor = comp.getClass().getAnnotation(ComponentActor.class);
    if (dontUpdateActor(syncActor)) {
      return;
    }
    updateActorComponent(entity, comp);
  }

  private void updateActorComponent(Entity entity, Component comp) {
    final ComponentChangedMessage msg = new ComponentChangedMessage();
    final ComponentMessageEnvelope envelope = new ComponentMessageEnvelope(entity.getId(), comp.getId(), msg);
    msgApi.sendToEntity(entity.getId(), envelope);
  }

  private boolean dontUpdateActor(ComponentActor syncActor) {
    return syncActor == null || !syncActor.updateActorOnChange();
  }
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/entity/component/interceptor/ActorUpdateComponentInterceptor.kt
package net.bestia.zoneserver.entity.component.interceptor

import net.bestia.entity.component.interceptor.BaseComponentInterceptor
import net.bestia.zoneserver.entity.Entity
import net.bestia.zoneserver.entity.EntityService
import net.bestia.zoneserver.entity.component.Component
import net.bestia.messages.entity.ComponentIdEnvelope
import net.bestia.messages.entity.EntityEnvelope

/**
 * This interceptor will check if the component was annotated to notify certain actors
 * if the component value has changed. The actor will then get notified.
 */
class ActorUpdateComponentInterceptor(
        private val msgApi: MessageApi
) : BaseComponentInterceptor<Component>(Component::class.java) {

  override fun onDeleteAction(entityService: EntityService, entity: Entity, comp: Component) {
    val syncActor = comp.javaClass.getAnnotation(ActorSync::class.java)
    if (dontUpdateActor(syncActor)) {
      return
    }
    updateActorComponent(entity, ComponentDeletedMessage(comp.id))
  }

  override fun onUpdateAction(entityService: EntityService, entity: Entity, comp: Component) {
    val syncActor = comp.javaClass.getAnnotation(ActorSync::class.java)
    if (dontUpdateActor(syncActor)) {
      return
    }
    updateActorComponent(entity, ComponentUpdateMessage(comp.id))
  }

  override fun onCreateAction(entityService: EntityService, entity: Entity, comp: Component) {
    val syncActor = comp.javaClass.getAnnotation(ActorSync::class.java)
    if (dontUpdateActor(syncActor)) {
      return
    }
    updateActorComponent(entity, ComponentCreatedMessage(comp.id))
  }

  private fun updateActorComponent(entity: Entity, msg: ComponentChangedMessage) {
    val compEnv = ComponentIdEnvelope(msg.componentId, msg)
    val entityEnv = EntityEnvelope(entity.id, compEnv)
    msgApi.send(entityEnv)
  }

  private fun dontUpdateActor(syncActor: ActorSync?): Boolean {
    return syncActor == null || !syncActor.updateActorOnChange
  }
}
